mlflow/tracing/export/mlflow.py

```diff
@@ -70,24 +76,159 @@ def export(self, root_spans: Sequence[ReadableSpan]):
+
+
+class Task:
+    """A dataclass to represent a task to be processed by the async trace export queue."""
+
+    def __init__(self, handler: Callable, args: Sequence, error_msg: str = ""):
+        self._handler = handler
+        self._args = args
+        self._error_msg = error_msg
+
+    def handle(self):
+        """Handle the task. This method must not raise any exception."""
         try:
-            self._client._upload_trace_spans_as_tag(trace.info, trace.data)
+            self._handler(*self._args)
         except Exception as e:
-            _logger.debug(f"Failed to log trace spans as tag to MLflow backend: {e}", exc_info=True)
-
-        # TODO: Make this async
-        # The trace is already updated in processor.on_end method
-        # so we just log to backend store here
-        try:
-            self._client._upload_trace_data(trace.info, trace.data)
-            self._client._upload_ended_trace_info(trace.info)
-        except Exception as e:
-            # avoid silent failures
             _logger.warning(
-                f"Failed to log trace to MLflow backend: {e}",
+                f"{self._error_msg} Error: {e}. For full traceback, set logging level to debug.",
                 exc_info=_logger.isEnabledFor(logging.DEBUG),
             )
+
+
+class AsyncTraceExportQueue:
+    """
+    This is a queue based run data processor that queue incoming data and process it using a single
+    worker thread. This class is used to process traces saving in async fashion.
+    """
+
+    def __init__(self, client) -> None:
+        self._queue: Queue[Task] = Queue()
+        self._client = client
+        self._lock = threading.RLock()
+
+        # A thread event that indicates the logging queue stop processing task.
+        self._stop_event = threading.Event()
+        self._is_activated = False
+        self._unprocessed_tasks = set()
+
+    def put(self, task: Task) -> None:
+        """Put a new task to the queue for processing."""
+        if not self.is_active():
+            self.activate()
+
+        # If stop even is set, we should wait for the queue to be drained before putting the task.
+        if self._stop_event.is_set():
+            self._stop_event.wait()
+
+        self._unprocessed_tasks.add(task)
+        self._queue.put(task)
+
+    def _logging_loop(self) -> None:
+        """
+        Continuously process incoming tasks in the queue until the stop event is set.
+        When the stop event is set, the loop will drain the remaining tasks in the queue.
+        """
+        while not self._stop_event.is_set():
+            self._handle_task()
+        while not self._queue.empty() or self._unprocessed_tasks:
+            self._handle_task()
+
+    def _handle_task(self) -> None:
+        """Process the given task in the running runs queues."""
+        try:
+            task = self._queue.get(timeout=1)
+        except Empty:
+            return
+
+        def _handle(task):
+            task.handle()
```

Comment:
If an Exception is thrown in this thread, this will bubble up and potentially terminate the main worker thread, right? Should we log exceptions and wrap the task.handle() in a try / catch / finally (within finally we remove from the _unprocessed_tasks set)?

Reply:
The exception is handled in the task's `handle()` method i.e. it should not raise any exception as I commented. I did so to allow different exception per task, but on second thought it is simpler to handle in this loop. Will update!

Reply:
Wait, on second thought, maybe it's still beneficial to have try-catch within `handle()` method. It allows us to avoid repeating try-catch for sync and async code path.